firedrake/output.py

```diff
@@ -651,3 +756,366 @@ def write(self, *functions, **kwargs):
+        # Construct function
+        f = Function(V, name=name)
+        tf = f.topological
+        tV = tf.function_space()
+        dm = tV._mesh.topology_dm
+        sectiondm = get_sectiondm_for_checkpointing(tV)
+        path = os.path.join("/topologies",
+                            dm.getName(),
+                            "dms",
+                            sectiondm.getName(),
+                            "vecs",
+                            tf.name())
+        # -- Embed if necessary
+        if self.get_attr(path, PREFIX + "_embedded"):
+            _f = self.load_function(PREFIX_EMBEDDED + f.name(), mesh_name, distribution_parameters=distribution_parameters)
+            element = V.ufl_element()
+            _element, method = get_embedding_element(element)
+            assert _element == _f.function_space().ufl_element()
+            if method == "project":
+                getattr(f, method)(_f, solver_parameters={"ksp_rtol": 1.e-16})
+            else:
+                getattr(f, method)(_f)
+            return f
+        vec = PETSc.Vec().createWithArray(tf.dat._data, size=None, bsize=tf.dat.cdim, comm=dm.comm)
+        vec.setName(tf.name())
+        dist_key = hash(frozenset(distribution_parameters.items()))
+        mesh_key = (mesh_name, dist_key)
+        sd_key = get_shared_data_key_for_checkpointing(tV.mesh(), tV.ufl_element())
+        _, _, sf = self._function_load_utils[sd_key + mesh_key]
+        dm.localVectorLoad(self.viewer, sectiondm, sf, vec)
+        return f
+
+    def find_sectiondm_name(self, mesh_name, fs_name):
+        base_path = os.path.join("/topologies", mesh_name, "dms")
+        for sectiondm_name in self.h5pyfile[base_path]:
+            path = os.path.join(base_path, sectiondm_name, PREFIX + "_function_spaces")
+            if fs_name in self.h5pyfile[path]:
+                return sectiondm_name
+        raise ValueError(f"FunctionSpace {fs_name} not found.")
+
+    def find_function_space_name(self, mesh_name, function_name):
+        if PREFIX + "_mixed_function_spaces" in self.h5pyfile["/topologies"]:
+            base_path = os.path.join("/topologies",
+                                     PREFIX + "_mixed_function_spaces")
+            for mixed_fs_name in self.h5pyfile[base_path]:
+                path = os.path.join(base_path, mixed_fs_name, PREFIX + "_functions")
+                self.h5pyfile.require_group(path)
+                if function_name in self.h5pyfile[path]:
+                    return mixed_fs_name
+        base_path = os.path.join("/topologies", mesh_name, "dms")
+        for sectiondm_name in self.h5pyfile[base_path]:
+            path = os.path.join(base_path, sectiondm_name, "vecs")
+            self.h5pyfile.require_group(path)
+            if function_name in self.h5pyfile[path]:
+                return self.get_attr(os.path.join(path, function_name), PREFIX + "_function_space")
+        raise ValueError(f"Function {function_name} not found.")
+
+    @property
```

Comment:
This is somewhat strange, because `cached_property` literally automates the exact code you have written here. Can you isolate the hang/reproduce outside a test suite?

Reply:
Yes. The following (almost always) hangs with `cached_property` with `mpiexec -n 2 ...`, but never with the current code as far as I've observed.
```
from firedrake import *
from pyop2.mpi import COMM_WORLD


filename = "test_hang.h5"
mesh_name = "default"
func_name = "f"


def _load_check_save_functions():
    with CheckpointFile(filename, "r") as afile:
        fB = afile.load_function(func_name, mesh_name=mesh_name)
    print(assemble(inner(fB, fB) * dx), flush=True)
    with CheckpointFile(filename, "w") as afile:
        afile.save_function(fB)
    print(f"rank = {COMM_WORLD.rank}: end of function:", flush=True)


def test_io_function_base():
    meshA = UnitSquareMesh(1, 1, quadrilateral=True)
    VA = FunctionSpace(meshA, "RTCF", 5, name="V")
    fA = Function(VA, name=func_name)
    with CheckpointFile(filename, 'w') as afile:
        afile.save_function(fA)
    for i in range(2):
        _load_check_save_functions()


test_io_function_base()
```
This looks very touchy to me. Even if it fails, if we run it again without doing `firedrake-clean`, it passes.